pallets/gear-program/src/migrations/v7.rs

```diff
@@ -0,0 +1,292 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{Config, Pallet, ProgramStorage};
+use common::{Origin, Program};
+use frame_support::{
+    traits::{tokens::Pay, Currency, Get, GetStorageVersion, OnRuntimeUpgrade, StorageVersion},
+    weights::Weight,
+};
+#[cfg(feature = "try-runtime")]
+use pallet_balances::Pallet as Balances;
+use pallet_balances::WeightInfo;
+#[cfg(feature = "try-runtime")]
+use sp_runtime::traits::UniqueSaturatedInto;
+use sp_std::marker::PhantomData;
+
+#[cfg(feature = "try-runtime")]
+use {
+    frame_support::ensure,
+    sp_runtime::{
+        codec::{Decode, Encode},
+        TryRuntimeError,
+    },
+    sp_std::vec::Vec,
+};
+
+const MIGRATE_FROM_VERSION: u16 = 6;
+const MIGRATE_TO_VERSION: u16 = 7;
+const ALLOWED_CURRENT_STORAGE_VERSION: u16 = 7;
+
+pub(crate) type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+pub(crate) type CurrencyOf<T> = <T as pallet_treasury::Config>::Currency;
+pub(crate) type BalanceOf<T> = <CurrencyOf<T> as Currency<AccountIdOf<T>>>::Balance;
+
+pub struct MigrateToV7<T: Config>(PhantomData<T>);
+
+impl<T: Config> OnRuntimeUpgrade for MigrateToV7<T>
+where
+    T: pallet_treasury::Config + pallet_balances::Config,
+    T::AccountId: Origin,
+    T::Paymaster: Pay<Beneficiary = T::AccountId, AssetKind = (), Balance = BalanceOf<T>>,
+{
+    fn on_runtime_upgrade() -> Weight {
+        let onchain = Pallet::<T>::on_chain_storage_version();
+        let existential_deposit = CurrencyOf::<T>::minimum_balance();
+        let transfer_weight = <T as pallet_balances::Config>::WeightInfo::transfer_allow_death();
+
+        // 1 read for on chain storage version
+        let mut weight = T::DbWeight::get().reads(1);
+
+        if onchain == MIGRATE_FROM_VERSION {
+            let current = Pallet::<T>::current_storage_version();
+
+            if current != ALLOWED_CURRENT_STORAGE_VERSION {
+                log::error!("❌ Migration is not allowed for current storage version {current:?}.");
+                return weight;
+            }
+
+            let update_to = StorageVersion::new(MIGRATE_TO_VERSION);
+            log::info!("🚚 Running migration from {onchain:?} to {update_to:?}, current storage version is {current:?}.");
+
+            let mut transferred = 0_u32;
+            let mut active_count = 0_u32;
+            ProgramStorage::<T>::iter().for_each(|(program_id, program)| {
+                weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
+
+                if let Program::Active(_p) = program {
+                    active_count += 1;
+                    // Transfer ED from treasury account to program's account
+                    // regardless whether the latter is extant or not.
+                    match <T as pallet_treasury::Config>::Paymaster::pay(
```

Comment:
Do we need to topup treasury before the upgrade?

Reply:
You're right. That's why they topped it up manually by 1k Vara a few days ago.